draft-ietf-mls-protocol.md

```diff
@@ -1867,6 +2019,42 @@ A member of the group applies a Remove message by taking the following steps:
 
 * Blank the intermediate nodes along the path from the removed leaf to the root
 
+### PSK
+
+A PSK proposal can be used to request that a PSK be injected into the
+key schedule in the process of advancing the epoch.
+
+~~~~~
+struct {
+    PreSharedKeyID psk;
+} PSK;
+~~~~~
+
+When processing a commit message including one or more PSK proposals, a group
+member MUST derive the `psk_secret` for the inclusion in the Key Schedule as
+described in {{pre-shared-keys}}, where the order of the PSKs corresponds to the
+order of the PSK proposals in the commit.
+
+### ReInit
```

Comment:
Honestly, I'm not sure how much value this adds.  The new group already has a pointer to the old group (by way of the PSK ID).  The best that this can do is cause folks not to use the old group any more, and it seems like doing that at the application layer is sufficient.  The only thing doing it in MLS adds is that it gets into the transcript -- and by definition, the transcript doesn't matter after this!

If we're going to have a message here, I would propose we do it a bit differently:

* Use a different message type, parallel to Proposal or Commit.  Commit advances the group to a new epoch; this operation does not.
* ... and explain that this needs the same sequencing as Commit.
* Have the semantic be just "close", with an optional forward pointer.
* For the forward pointer, just use a hash of the Welcome message.

```
opaque WelcomeID<0..255>; 

struct {
    optional<WelcomeID> welcome_hash;
} Close;
```

Reply:
I understand that you're reluctant to bloat the proposal space like this. However, I think a ReInit proposal serves two purposes. 1) It allows an external party (e.g. the DS) to indicate to the group that they should change/upgrade the group. A ReInit proposal authenticates the DS' intent and even the upgrade/change it wants to do.
2) It makes sure that the group agrees on the ReInit. The PSK in the Welcome message is derived from the epoch _after_ the commit, so everyone must have processed it and registered that the group should be closed.

I think both authentication and agreement are valuable properties for an action that has the power to change ciphersuite, version and extensions.

Other than potentially bloating the number of proposals we have: Why not have a ReInit proposal?

Reply:
> As to "why not" -- typically we try not to put things in the protocol that don't do anything ;)

Point taken  :smile: 

Regarding the handshake message: I wasn't aware that they're also pretty extensible, so sure, that would work as well. Just to make sure I understand correctly: The server would be able to send the message themselves (if they're "preconfigured") and the message would be authenticated, right? Also, we would have to add the data fields to the handshake message instead like so:
```
struct {
    opaque group_id<0..255>;
    uint64 epoch;
    Sender sender;
    opaque authenticated_data<0..2^32-1>;

    ContentType content_type;
    select (MLSPlaintext.content_type) {
        case application:
          opaque application_data<0..2^32-1>;

        case proposal:
          Proposal proposal;

        case commit:
          Commit commit;
          opaque confirmation_tag<0..255>;

        case close:
          ProtocolVersion version;
          CipherSuite cipher_suite;
          Extension extensions<0..2^32-1>;
    }

    opaque signature<0..2^16-1>;
} MLSPlaintext;
```
For maximum security guarantees, we could also add a `confirmation_tag` just like for the commit.

Reply:
Very good points. Using the Welcome message seems like an elegant way to link the groups.
Regarding ReInit vs. Replace, I have no strong opinion on that. Replace seems a little more direct, so I'll go with that for now. I'll make the changes either today or over the weekend so hopefully we can get the green light for a merge on Tuesday :-)

Reply:
I just noticed that if we use a Welcome message (or its hash) in the Replace message, then the Server won't be able to send it, because it lacks the necessary information to craft it. So until we find a better way of doing it, I'm going with the way I originally sketched. Let me know if you have another idea.

Reply:
If we go for a Handshake Message as opposed to a Proposal, it's also not clear, which group member sends the Welcome message for the new group, as the Close message can come from the Server itself. The Server could nominate a member, but it doesn't know if that member is online and so there is a risk that other members would then have to wait for that member to come online. The other possibility is to just have any member do it, which would create a race condition and possibly multiple follow-up groups.

 Do you have an idea on how to deal with that? Since there is otherwise not that much of a difference between Handshake Message and Proposal, I'm inclined to stick with a Proposal, but rename it to "Close" and make the content optional.